src/pkcs11/core_pkcs11_test.c

```diff
@@ -0,0 +1,2647 @@
+    CK_SESSION_HANDLE xSession;
+
+    memcpy( &xSession, pxMultiTaskParam->pvTaskData, sizeof( CK_SESSION_HANDLE ) );
+
+    for( xCount = 0; xCount < PKCS11_TEST_MULTI_THREAD_LOOP_COUNT; xCount++ )
+    {
+        xResult = pxGlobalFunctionList->C_GenerateRandom( xSession,
+                                                          xRandomData,
+                                                          sizeof( xRandomData ) );
+
+        if( xResult != CKR_OK )
+        {
+            TEST_PRINTF( "GenerateRandom multi-thread task failed.  Error: %ld \r\n", xResult );
+            break;
+        }
+    }
+
+    /* Report the result of the loop. */
+    pxMultiTaskParam->xTestResult = xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+TEST( Full_PKCS11_NoObject, AFQP_GenerateRandomMultiThread )
+{
+    uint32_t xTaskNumber;
+    CK_SESSION_HANDLE xSessionHandle[ PKCS11_TEST_MULTI_THREAD_TASK_COUNT ];
+
+    for( xTaskNumber = 0; xTaskNumber < PKCS11_TEST_MULTI_THREAD_TASK_COUNT; xTaskNumber++ )
+    {
+        xInitializePkcs11Session( &xSessionHandle[ xTaskNumber ] );
+        xGlobalTaskParams[ xTaskNumber ].pvTaskData = &xSessionHandle[ xTaskNumber ];
+    }
+
+    prvMultiThreadHelper( ( void * ) prvGenerateRandomMultiThreadTask );
+
+    for( xTaskNumber = 0; xTaskNumber < PKCS11_TEST_MULTI_THREAD_TASK_COUNT; xTaskNumber++ )
+    {
+        pxGlobalFunctionList->C_CloseSession( xSessionHandle[ xTaskNumber ] );
+    }
+}
+
+/*--------------------------------------------------------*/
+/*-------------- RSA Tests ------------------------------ */
+/*--------------------------------------------------------*/
+
+TEST_SETUP( Full_PKCS11_RSA )
+{
+    CK_RV xResult;
+
+    xResult = xInitializePKCS11();
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to initialize PKCS #11 module." );
+
+    xResult = xInitializePkcs11Session( &xGlobalSession );
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to open PKCS #11 session." );
+}
+
+/*-----------------------------------------------------------*/
+
+TEST_TEAR_DOWN( Full_PKCS11_RSA )
+{
+    CK_RV xResult;
+
+    xResult = pxGlobalFunctionList->C_CloseSession( xGlobalSession );
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to close session." );
+
+    xResult = pxGlobalFunctionList->C_Finalize( NULL );
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to finalize session." );
+}
+
+/*-----------------------------------------------------------*/
+
+TEST_GROUP_RUNNER( Full_PKCS11_RSA )
+{
+    #if ( PKCS11_TEST_RSA_KEY_SUPPORT == 1 )
+        /* Reset the cryptoki to uninitialize state. */
+        prvBeforeRunningTests();
+
+        #if ( PKCS11_TEST_IMPORT_PRIVATE_KEY_SUPPORT == 1 )
+            RUN_TEST_CASE( Full_PKCS11_RSA, AFQP_CreateObject );
+        #endif
+
+        /* Generating RSA keys is not currently supported.
+         #if ( PKCS11_TEST_GENERATE_KEYPAIR_SUPPORT == 1 )

```

Comment:
Add a` #error` at the beginning of file, saying that this option is not supported, will help users to identify any issues if they have this option enabled, rather than failing silently.

Reply:
RSA keypair generation is currently not supported in corePKCS mbedtls implementation.